consensus/core/src/core.rs

```diff
@@ -442,33 +440,71 @@ impl Core {
             .start_timer();
 
+        let mut committed_subdags = Vec::new();
         // TODO: Add optimization to abort early without quorum for a round.
-        let sequenced_leaders = self.committer.try_commit(self.last_decided_leader);
+        loop {
+            // LeaderSchedule has a limit to how many sequenced leaders can be committed
+            // before a change is triggered. Calling into leader schedule will get you
+            // how many commits till next leader change. We will loop back and recalculate
+            // any discarded leaders with the new schedule.
+            let sequenced_leaders = self
+                .committer
+                .try_commit(self.last_decided_leader)
+                .into_iter()
+                .take(
+                    self.leader_schedule
+                        .commits_until_leader_schedule_update(self.dag_state.clone()),
+                )
+                .collect::<Vec<_>>();
 
-        if let Some(last) = sequenced_leaders.last() {
-            self.last_decided_leader = last.get_decided_slot();
-            self.context
-                .metrics
-                .node_metrics
-                .last_decided_leader_round
-                .set(self.last_decided_leader.round as i64);
-        }
+            if sequenced_leaders.is_empty() {
+                break;
+            }
 
-        let committed_leaders = sequenced_leaders
-            .into_iter()
-            .filter_map(|leader| leader.into_committed_block())
-            .collect::<Vec<_>>();
-        if !committed_leaders.is_empty() {
-            debug!(
-                "Committing leaders: {}",
-                committed_leaders
-                    .iter()
-                    .map(|b| b.reference().to_string())
-                    .join(",")
-            );
+            if let Some(last) = sequenced_leaders.last() {
+                self.last_decided_leader = last.get_decided_slot();
+                self.context
+                    .metrics
+                    .node_metrics
+                    .last_decided_leader_round
+                    .set(self.last_decided_leader.round as i64);
+            }
+
+            let committed_leaders = sequenced_leaders
+                .into_iter()
+                .filter_map(|leader| leader.into_committed_block())
+                .collect::<Vec<_>>();
+            if !committed_leaders.is_empty() {
+                debug!(
+                    "Committing leaders: {}",
+                    committed_leaders
+                        .iter()
+                        .map(|b| b.reference().to_string())
+                        .join(",")
+                );
+            }
+
+            let subdags = self.commit_observer.handle_commit(committed_leaders)?;
+            self.dag_state
+                .write()
+                .add_unscored_committed_subdags(subdags.clone());
+            committed_subdags.extend(subdags);
+
+            if self
```

Comment:
`break` when the condition is false? Otherwise it will waste time try committing again.